### DISCUSSION TITLE
Detect and build source code

### DISCUSSION THREAD

**Discussion by RATANAJANGIR** [buildpacks/community#217] [General]
Hi, is there any method in build Pack to just detect source code and build it. I don't want to make image; I want to publish output of build phase (source code) (ex. .jar package, wheel package).
Really appreciated for any help here.

Thanks,

**Comment by natalieparellano** [comment]
Hi @RATANAJANGIR `pack` doesn't currently support a workflow like that, but you might be able to get what you need by exercising the lifecycle directly (just the `detect` and `build` phases). Our (somewhat outdated) Tekton task template gives an idea of how to do that (see [here](https://github.com/tektoncd/catalog/blob/main/task/buildpacks-phases/0.2/buildpacks-phases.yaml#L142-L156) and [here](https://github.com/tektoncd/catalog/blob/main/task/buildpacks-phases/0.2/buildpacks-phases.yaml#L189-L204)).

**Reply by RATANAJANGIR** [reply]
Thanks @natalieparellano, can you please help to get output of **build** phase! I want to implement one other lifecycle which should take output of build phase and publish artifact to Jfrog registry, after that I can make merge request in your code.

**Reply by natalieparellano** [reply]
Nice :) the exporter is spec'd here: https://github.com/buildpacks/spec/blob/main/platform.md#outputs-5

Copied below are select responsibilities of the exporter that seem most relevant to your use case. But you'll definitely want to confirm that the expected functionality is present with any alternative implementation of the lifecycle:

> The app image:
> * MUST contain all buildpack-provided launch layers as determined by the [Buildpack Interface Specfication](https://github.com/buildpacks/spec/blob/main/buildpack.md)
> * MUST contain one or more app layers as determined by the [Buildpack Interface Specfication](https://github.com/buildpacks/spec/blob/main/buildpack.md)
> * MUST contain one or more launcher layers that include:
>   * A file with the contents of the `<launcher>` file at path `/cnb/lifecycle/launcher`
>   * One symlink per buildpack-provided process type with name `/cnb/process/<type>` and target `/cnb/lifecycle/launcher`
> * MUST contain a layer that includes `<layers>/config/metadata.toml`

**Reply by RATANAJANGIR** [reply]
@natalieparellano, is this possible to get runnable artifact  from "/cnb/lifecycle/builder" the command. 
I just want to use "/cnb/lifecycle/detector" and "/cnb/lifecycle/builder", I tried with python code, but not able to see  ".whl" package in building lifecycle


![image](https://github.com/buildpacks/community/assets/50768957/4ca5dd2b-f86b-46d2-902c-dbc72496dd76)

**Comment by natalieparellano** [comment]
@RATANAJANGIR by the time the build phase exits, the buildpack should have created whatever artifacts are necessary for the application to be runnable. The specific behavior will vary by buildpack - in this case, as you're using Paketo, I'd recommend checking out https://paketo.io/docs/howto/python/. You can inspect the `/layers` directory of your output application image to learn more about how the buildpacks package the application.

**Reply by RATANAJANGIR** [reply]
Hi @natalieparellano , Thanks for quick reply, I have one other question->

I am using packeto-buildpack/base-builder for builder.toml configration but for python image I am not using composite buildpack and I wrote one buildpack for python which contains buildpack.toml, bin/detect and bin/build and replaced URI of python buildpack in builder.toml, 

But I am getting some error when running "pack builder create builder-image --config builder.toml" 


![image](https://github.com/buildpacks/community/assets/50768957/ed16110c-fa57-4dbc-8bc8-c9a85e913fdf)


ERROR: validating buildpacks: buildpack buildpack/python@1.0.0 does not support stack io.buildpacks.stacks.bionic

For python I have added buildpack in builder.toml
```
[[buildpacks]]
  uri = " harbor-preprod.blackrock.com/data-ops-engineering/chapters/devops-engineering/python-buildpack:1.0.0-debian-bullseye"
  version = "1.0.0"

[[order]]

  [[order.group]]
    id = "buildpack/python"
    version = "1.0.0"
```

I need help to understand this stack concept with BuildPack.

**Comment by natalieparellano** [comment]
Ah - to fix this error you need to declare 

```
[[stacks]]
  id = "io.buildpacks.stacks.bionic"
```

in your buildpack.toml (not builder.toml) for the custom buildpack. This allows the buildpack to declare which build/run base image combinations it is compatible with. 

I should note that the concept of "stack" is [going away](https://github.com/buildpacks/rfcs/blob/main/text/0096-remove-stacks-mixins.md) in the next set of APIs. Instead of declaring compatibility with a magical ID such as `io.buildpacks.stacks.bionic` a buildpack can declare that it works with `os = linux` `arch = amd64` and so on (more details in the linked RFC).

**Reply by RATANAJANGIR** [reply]
Thanks @natalieparellano, I have one question, in bin/build I am running this command 

```
 #!/usr/bin/env bash

set -euo pipefail

PIP_LAYER="$CNB_LAYERS_DIR/pip"
mkdir -p "$PIP_LAYER/modules" "$PIP_LAYER/env"

python3 setup.py bdist_wheel

echo "launch = true" > "$PIP_LAYER.toml"

```
Can you please tell how to setup python here to run  "python3 setup.py bdist_wheel" command

**Reply by RATANAJANGIR** [reply]
I am installing python in container image but its giving unexpected error
![image](https://github.com/buildpacks/community/assets/50768957/2a5c9a44-3679-459f-b150-20c54a4a6335)



there is any other way to install python like if I am using my custom-buildpack, So how I can use dependencies so that before running "python3 setup.py bdist_wheel" command it will setup python in ENV var, Like Paketo-buildpack/python.

**Reply by natalieparellano** [reply]
Alas, this is nearing the limits of my expertise :) I would recommend reaching out in the #python channel in slack.paketo.io or #buildpack-authors channel in slack.cncf.io for more specific guidance related to Python.
